Refactor kube-scheduler config API, command, and server setup

ironcladlou opened the pull request:
Refactor the kube-scheduler configuration API, command setup, and server setup according to the guidelines established in #32215 and using the kube-proxy refactor (#34727) as a model of a well factored component adhering to said guidelines.

* Config API: clarify meaning and use of algorithm source by replacing modality derived from bools and string emptiness checks with an explicit AlgorithmSource type hierarchy.
* Config API: consolidate client connection config with common structs.
* Config API: split and simplify healthz/metrics server configuration.
* Config API: clarify leader election configuration.
* Config API: improve defaulting.
* CLI: deprecate all flags except `--config`.
* CLI: port all flags to new config API.
* CLI: refactor to match kube-proxy Cobra command style.
* Server: refactor away configurator.go to clarify application wiring.
* Server: refactor to more clearly separate wiring/setup from running.

Fixes https://github.com/kubernetes/kubernetes/issues/52428.

@kubernetes/api-reviewers 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@kubernetes/sig-scheduling-pr-reviews 

/cc @ncdc @timothysc @bsalamat

```release-note
The kube-scheduler command now supports a `--config` flag which is the location of a file containing a serialized scheduler configuration. Most other kube-scheduler flags are now deprecated.
```

timothysc commented:
@ironcladlou Just ping me when you want me folks to review.

ironcladlou commented:
/test pull-kubernetes-unit

ironcladlou commented:
@timothysc I'd love to get some initial feedback before I go any further with testing and polishing.

ironcladlou commented:
/test pull-kubernetes-unit

ironcladlou commented:
Before I spend time doing stuff like fixing tests and (possibly) moving the config API to another API group, I'd like to get some feedback on the proposed API and other refactorings.

ironcladlou commented:
Hi. My name is Dan, and I'm a rebase survivor. 😐

ironcladlou commented:
@timothysc ping, any chance you could take a look here?

mikedanese commented:
for my parts:
/approve

bsalamat commented:
I took a quick look and generally like the idea and direction of this PR, particularly I like the more streamlined scheduler configuration method and cleaner configuration types, but the PR is hard to review because various changes, including addition of new types for scheduler config, auto-generated files, and changes to the initialization of scheduler and their tests are all combined in a single commit.
It would be great if you could separate them in smaller commits.

ironcladlou commented:
@bsalamat thanks for the feedback. I split the changes into three bite-sized commits, and will continue forward with tests and such while awaiting more detailed review of the existing changes.

timothysc commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -120,41 +120,70 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
+			obj.AlgorithmSource.Provider = &val
+		}
+	case SchedulerAlgorithmSourcePolicy:
+		// There is no default file or config map to apply.
```
@bsalamat ^ Should we choose one?  The longstanding plan was to default to configmap.

ironcladlou replied:
We do support a default namespace for the config map, but not a default name. A default name sounds good to me. Perhaps somebody else can think of a reason we shouldn't.

bsalamat replied:
When there is no policy file or policy ConfigMap, scheduler uses its builtin policy. We may change that one day, but that is beyond the scope of this PR.

timothysc reviewed (commented):
Sorry for the delay, this was on my todo, but I haven't had time till recently.  

So generally looks ok, but I have logistical questions.  
e.g. -  What will be the flag deprecation plan?  

Also please squash to 2 commits.. 1 code changes and 2 generated changes.

ironcladlou commented:
@timothysc 
>Sorry for the delay, this was on my todo, but I haven't had time till recently.

Probably a good thing, since it was just today I was able to get back to it and really start finishing the PR to get it out of WIP status...

>So generally looks ok, but I have logistical questions.
>e.g. - What will be the flag deprecation plan?

Good question, and I'm not sure. I'm closely following @ncdc's proxy refactor here. Would the answer for the proxy flags be the same here?

ncdc commented:
Yes. See https://kubernetes.io/docs/reference/deprecation-policy/, specifically "Deprecating a flag or CLI". I'm assuming we'll fall under "1 year or 2 releases (whichever is longer)". Also, that doc doesn't spell out how to announce a deprecation. For the kube-proxy config, I added a release note (https://github.com/kubernetes/kubernetes/blob/6c081c3d53ad043f43ac3d5f976a83431f2a9cb4/CHANGELOG.md#kube-proxy).

1 possible wrinkle (and this probably isn't really the right place for discussion) is how do we all feel about deprecating "stable" flags and replacing them with what is currently an alpha API? I suppose if we commit to graduating the API go GA by the time the deprecated flags are removed, that might be an acceptable solution.

Also, completely unrelated, @ironcladlou you should move the scheduler config structs to their own API group, so you don't have to do that in a follow-up (see e.g. #53645)

deads2k commented on pkg/apis/componentconfig/types.go:
```diff
@@ -160,58 +161,115 @@ const (
 
+// SchedulerPolicyConfigMapKey defines the key of the element in the
+// scheduler's policy ConfigMap that contains scheduler's policy config.
+const SchedulerPolicyConfigMapKey string = "policy.cfg"
```
pre-existing?  I would have expected an extension like .yaml or something.

ironcladlou replied:
Yeah, this is left unchanged.

bsalamat replied:
> I would have expected an extension like .yaml or something.

Please note that this is not a file name. So, the .cfg is not really an extension that implies the format of the content. It is merely a key to the ConfigMap.

deads2k commented on pkg/apis/componentconfig/types.go:
```diff
@@ -160,58 +161,115 @@ const (
+// mutually exclusive.
+type SchedulerPolicySource struct {
+	// Type is the policy source type.
+	Type SchedulerPolicySourceType
```
unnecessary if you have a rule about "only one is non-nil", right?

ironcladlou replied:
Yeah. Is omitting a "type" in a struct like this with mutually exclusive fields the preferred style?

ironcladlou replied:
Removed Type.

deads2k commented on pkg/apis/componentconfig/types.go:
```diff
@@ -160,58 +161,115 @@ const (
+type SchedulerPolicyConfigMapSource struct {
+	// Reference is a config map containing a seralized policy in the
+	// SchedulerPolicyConfigMapKey key.
+	Reference api.ObjectReference
```
I think we've agreed not to depend on this since the type is overly large and actually wrong about depending on things like kind vs. resources.  Specific types for specific cases as I recall brian's advice, right @liggitt ?

liggitt replied:
yeah, use-specific object

ironcladlou replied:
I removed the use of ObjectReference.

deads2k commented on pkg/apis/componentconfig/types.go:
```diff
@@ -160,58 +161,115 @@ const (
+type SchedulerAlgorithmSource struct {
+	Type SchedulerAlgorithmSourceType
+	// Policy is a policy based algorithm source.
+	Policy *SchedulerPolicySource
```
is this and provider mutually exclusive?

ironcladlou replied:
Correct, the specific source fields are mutually exclusive.

ironcladlou replied:
Removed Type.

deads2k commented on pkg/apis/componentconfig/types.go:
```diff
@@ -160,58 +161,115 @@ const (
+type KubeSchedulerLeaderElectionConfiguration struct {
+	LeaderElectionConfiguration
 	// LockObjectNamespace defines the namespace of the lock object
 	LockObjectNamespace string
```
Seems like this is a thing we'd want in general for leader election. The controllers logically need it for subdivision.

ironcladlou replied:
Fine with me. Currently the only other LeaderElectionConfiguration consumer I found (KubeControllerManagerConfiguration) uses a hard-coded lock namespace/name. Could move these fields to LeaderElectionConfiguration and refactor the controller manager later.

guangxuli commented:
/cc

ironcladlou commented:
This is still ready for more review. One major outstanding question is whether we want the types put into their own API group with this PR (https://github.com/kubernetes/kubernetes/pull/52562#issuecomment-335901103).

ironcladlou commented:
@bsalamat @timothysc @ncdc @deads2k PTAL. I'm a little worried about dragging this PR on even longer by moving the API to its own group, especially with the frequent rebasing I've had to keep up with. Can the API group refactor be a followup?

timothysc commented:
@ironcladlou let us know when you want us to do a formal pass-through.  
/cc @bsalamat

bsalamat commented:
This is in my todo list. I will try to review this tomorrow.

ironcladlou commented:
I removed the SchedulerAlgorithmSourceType and SchedulerPolicySourceType fields (per https://github.com/kubernetes/kubernetes/pull/52562#discussion_r144842908 and https://github.com/kubernetes/kubernetes/pull/52562#discussion_r144842341).

bsalamat commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -129,18 +129,29 @@ func SetDefaults_KubeSchedulerConfiguration(obj *KubeSchedulerConfiguration) {
+	}
+
+	if obj.AlgorithmSource.Policy != nil {
+		switch obj.AlgorithmSource.Policy.Type {
```
Why do we need this switch case? All that is happening here is to set a default namespace for a provided ConfigMap when it doesn't have a namespace.

bsalamat commented on test/integration/scheduler/scheduler_test.go:
```diff
@@ -82,181 +79,6 @@ func PriorityTwo(pod *v1.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo
 	return []schedulerapi.HostPriority{}, nil
 }
 
-// TestSchedulerCreationFromConfigMap verifies that scheduler can be created
```
Why is this test deleted? We need a test on creating scheduler from a ConfigMap.

ironcladlou replied:
Creating a scheduler from a serialized policy is already covered in the [TestCreateFromConfig  unit test](https://github.com/ironcladlou/kubernetes/blob/kube-scheduler-config/plugin/pkg/scheduler/factory/factory_test.go#L76). A new additional unit test to cover [makeSchedulerConfig](https://github.com/ironcladlou/kubernetes/blob/kube-scheduler-config/plugin/cmd/kube-scheduler/app/server.go#L647) would also be useful to ensure we actually attempt to follow the code paths in TestCreateFromConfig when set up to use the config map source.

The integration test relies on assertions against internal state of the scheduler exposed for the sole purpose of this particular integration test.

Switching to unit tests in this case could provide comparable coverage but faster and without exposing state unnecessarily.

Happy to try and restore this if you're unconvinced, please let me know what you think.

bsalamat replied:
The integration test was also testing the fact that scheduler reads the config from etcd and parses it. This functionality is not tested in the unit test.

ironcladlou replied:
The two configmap integration tests have been restored.

bsalamat commented on plugin/cmd/kube-scheduler/app/server.go:
```diff
@@ -59,34 +334,169 @@ constraints, affinity and anti-affinity specifications, data locality, inter-wor
+}
+
+// NewSchedulerServer creates a runnable SchedulerServer from configuration.
+func NewSchedulerServer(config *componentconfig.KubeSchedulerConfiguration, master string) (*SchedulerServer, error) {
```
Please break this function into smaller functions. It is too long.

ironcladlou replied:
Decomposed.

bsalamat commented on plugin/cmd/kube-scheduler/app/server.go:
```diff
@@ -59,34 +334,169 @@ constraints, affinity and anti-affinity specifications, data locality, inter-wor
-		s,
-		kubeClient,
+// Run runs the SchedulerServer. This should never exit.
+func (s *SchedulerServer) Run() error {
```
Same as above. Please break this into smaller functions.

ironcladlou replied:
Decomposed.

bsalamat reviewed (commented):
Thanks, @ironcladlou!
I did a first pass and changes look fine to me overall.

Next time, please commit autogenerated changes under a separate commit.

Also, after you address reviewers comments, commit your changes under a new commit to make the review process easier.

ironcladlou commented:
/retest

ironcladlou commented:
/test pull-kubernetes-unit

bsalamat commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -120,41 +120,56 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
+
+	if len(obj.HealthzBindAddress) == 0 {
+		obj.HealthzBindAddress = fmt.Sprintf("0.0.0.0:%v", ports.SchedulerPort)
+	} else if !strings.Contains(obj.HealthzBindAddress, ":") {
```
This check and similar ones below are going to cause IPv6 compatibility issues. Please use [go 'net' package](https://golang.org/pkg/net/) to add/find port in an address.

ironcladlou replied:
Done, thanks for the tip. Much cleaner, too.

bsalamat commented on plugin/cmd/kube-scheduler/app/server.go:
```diff
@@ -18,37 +18,307 @@ limitations under the License.
+	index := strings.Index(o.config.HealthzBindAddress, ":")
+	if index != -1 {
+		port := o.config.HealthzBindAddress[:index+1]
+		bindAddr := fmt.Sprintf("%s:%s", o.healthzAddress, port)
```
Please use [net.joinHostPort](https://golang.org/pkg/net/#JoinHostPort) for IPv6 compatibility.

ironcladlou replied:
Fixed

bsalamat commented on plugin/cmd/kube-scheduler/app/server.go:
```diff
@@ -18,37 +18,307 @@ limitations under the License.
+	index := strings.Index(o.config.HealthzBindAddress, ":")
+	if index != -1 {
+		addr := o.config.HealthzBindAddress[:index]
+		bindAddr := fmt.Sprintf("%s:%d", addr, o.healthzPort)
```
Similar to other places. Please go net package.

ironcladlou replied:
Fixed

bsalamat reviewed (commented):
This PR is larger than I can confidently review, but I guess we don't have many other options. It looks generally good to me and seems our tests are fine. I have a couple of comments on IPv6 compatibility.

bsalamat commented:
Thanks, @ironcladlou! Looks good. Please go ahead and squash your commits, but leave the auto-generated files as a separate commit.

ironcladlou commented:
@bsalamat this is squashed.

bsalamat commented:
Thanks, @ironcladlou! Could you please add a release note?

ironcladlou commented:
Added a release note.

bsalamat commented:
/lgtm

xiangpengzhao commented:
#53645 merged. I guess it's a bad boy. It makes many PRs need rebase...

ironcladlou committed:
Update generated files

ironcladlou committed:
Refactor scheduler config API

Refactor the kube-scheduler configuration API, command setup, and server
setup according to the guidelines established in #32215 and using the
kube-proxy refactor (#34727) as a model of a well factored component
adhering to said guidelines.

* Config API: clarify meaning and use of algorithm source by replacing
modality derived from bools and string emptiness checks with an explicit
AlgorithmSource type hierarchy.
* Config API: consolidate client connection config with common structs.
* Config API: split and simplify healthz/metrics server configuration.
* Config API: clarify leader election configuration.
* Config API: improve defaulting.
* CLI: deprecate all flags except `--config`.
* CLI: port all flags to new config API.
* CLI: refactor to match kube-proxy Cobra command style.
* Server: refactor away configurator.go to clarify application wiring.
* Server: refactor to more clearly separate wiring/setup from running.

Fixes #52428.

ironcladlou commented:
Rebased. Will see if the tests survived.

bsalamat commented:
/lgtm

resouer commented:
@ironcladlou This change actually caused confusion recently, for example https://stackoverflow.com/questions/47966440/how-to-create-a-config-file-for-kube-scheduler-to-use-by-the-config-argument, also some users are asking this in slack.

I think maybe you can amend an example configure yaml for scheduler?
